### PR TITLE
Start partitions on corresponding assignments.stable, calculate if missing, cleanup obsolete resources (part 1)

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-20210

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 1]
```diff
@@ -1057,9 +1079,8 @@ private CompletableFuture<?> createTableLocally(long causalityToken, int catalog
 
             // Check if the table already has assignments in the vault.
             // So, it means, that it is a recovery process and we should use the vault assignments instead of calculation for the new ones.
-            // TODO: IGNITE-20210 Fix it
-            if (partitionAssignments(vaultManager, tableId, 0) != null) {
-                assignmentsFuture = completedFuture(tableAssignments(vaultManager, tableId, zoneDescriptor.partitions()));
+            if (partitionAssignments(metaStorageMgr, tableId, 0, causalityToken) != null) {
```
Comment (vldpyatkov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1082]: Why do you use the call for 0 participation only?

Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1082]: It's used only for condition to peek whether there are any assignments in meta storage. If yes, we read the assignments themselves below

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 2]
```diff
@@ -629,201 +630,222 @@ private CompletableFuture<?> createTablePartitionsLocally(
...
+                return ByteUtils.toBytes(assignments);
+            }
 
-                var safeTimeTracker = new PendingComparableValuesTracker<HybridTimestamp, Void>(
-                        new HybridTimestamp(1, 0)
-                );
-                var storageIndexTracker = new PendingComparableValuesTracker<Long, Void>(0L);
+            @Override
+            public long revision() {
+                return 1;
```
Comment (kgusakov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:686]: I would prefer to throw `java.lang.UnsupportedOperationException`here and in other unused methods - to simplify the debugging of any connected issues in future.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 3]
```diff
@@ -517,23 +516,26 @@ static CompletableFuture<List<Set<Assignment>>> tableAssignments(
     /**
      * Returns table assignments for all table partitions from vault.
```
Comment (kgusakov) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:517]: "from vault" the same as previous

**Comment by kgusakov** [comment]
As I understand: we are not cleaning resources from the obsolete partitions on recovery. So, if there were some updates for `stable` assignments during the node unavailability - we will have some resources leak. Is this issue out of ticket scope? Because now it is part of it's title

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 4]
```diff
@@ -466,17 +465,17 @@ public static CompletableFuture<Set<Assignment>> partitionAssignments(
     /**
      * Returns partition assignments from vault.
      *
-     * @param vaultManager Vault manager.
+     * @param metaStorageManager Meta storage manager.
      * @param tableId Table id.
      * @param partitionNumber Partition number.
+     * @param revision Revision.
      * @return Returns partition assignments from vault or {@code null} if assignments is absent.
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:472]: fixed

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java (hunk 3 above)
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:517]: fixed

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 5]
```diff
@@ -493,7 +493,7 @@ void testDestroyPartitionStoragesOnEvictNode() throws Exception {
     @Test
     @UseTestTxStateStorage
     @UseRocksMetaStorage
-    @Disabled("https://issues.apache.org/jira/browse/IGNITE-20210")
+    @Disabled("https://issues.apache.org/jira/browse/IGNITE-21463")
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:496]: Accidently put IGN ticket number, fixed

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 2 above)
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:686]: Fixed.

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 6]
```diff
@@ -690,11 +690,19 @@ private void directUpdateMetastoreRebalanceAssignmentKeys() throws Exception {
...
-        for (int i = 0; i < NODE_COUNT; i++) {
-            verify(getNode(i).raftManager, times(1))
-                    .startRaftGroupNode(any(), any(), any(), any(), any(RaftGroupOptions.class));
-            verify(getNode(i).replicaManager, times(1)).startReplica(any(), any(), any(), any(), any());
-        }
+        waitForCondition(() -> {
+            try {
+                for (int i = 0; i < NODE_COUNT; i++) {
+                    verify(getNode(i).raftManager, times(1))
+                            .startRaftGroupNode(any(), any(), any(), any(), any(RaftGroupOptions.class));
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java]: Reworked.

**Review by kgusakov — APPROVED** [review]

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java [hunk 7]
```diff
@@ -500,7 +500,7 @@ void testDestroyPartitionStoragesOnEvictNode() throws Exception {
     @Test
     @UseTestTxStateStorage
     @UseRocksMetaStorage
-    @Disabled("https://issues.apache.org/jira/browse/IGNITE-20210")
+    @Disabled("https://issues.apache.org/jira/browse/IGNITE-19170")
     void testDestroyPartitionStoragesOnRestartEvictedNode(TestInfo testInfo) throws Exception {
```
Comment (sanpwc) [inline modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:504]: Should we write some tests that will ensure that proposed implementation is correct?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 8]
```diff
@@ -523,38 +520,57 @@ public void start() {
     }
 
     private void performRebalanceOnRecovery(long recoveryRevision) {
-        CompletableFuture<Void> pendingAssignmentsRecoveryFuture;
-
-        var prefix = new ByteArray(PENDING_ASSIGNMENTS_PREFIX);
+        var pendingAssignmentsPrefix = new ByteArray(PENDING_ASSIGNMENTS_PREFIX);
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:523]: Why did you change performRebalanceOnRecovery? rebalance is about handling pending keys. I expect it to be properly implemented in IGNITE-20344. Did you find bugs there?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 9]
```diff
@@ -523,38 +520,57 @@ public void start() {
...
+        startVv.update(recoveryRevision, (v, e) -> handleAssignmentsEventsOnRecovery(
+                pendingAssignmentsPrefix,
+                recoveryRevision,
+                (entry, rev) -> handleChangePendingAssignmentEvent(entry, localPartsByTableIdVv.get(recoveryRevision), rev, true),
+                "pending"
+        ));
+    }
 
-        try (Cursor<Entry> cursor = metaStorageMgr.prefixLocally(prefix, recoveryRevision)) {
+    private CompletableFuture<Void> handleAssignmentsEventsOnRecovery(
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:540]: What do you mean? On recovery there are no events, right? We just scan catalog, retrive corresponding stable and pending assignments, etc? What kind of events do you handle on recovery?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 10]
```diff
@@ -630,201 +646,222 @@ private CompletableFuture<?> createTablePartitionsLocally(
...
-
             for (int i = 0; i < partitions; i++) {
                 int partId = i;
 
                 Set<Assignment> newPartAssignment = newAssignments.get(partId);
 
-                InternalTable internalTbl = table.internalTable();
+                // TODO https://issues.apache.org/jira/browse/IGNITE-19170 #handleChangePendingAssignmentEvent should be called on
+                // TODO actual event, the method #createTablePartitionsLocally should be removed.
+                handleChangePendingAssignmentEvent(
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:665]: What's the point in considering initial assignments as pending one? Generally it sounds reasonable but only if you took into consideration the fact that assignments.stable might be empty and in this case there's no sense in calling changePeersAsync, did you?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 11]
```diff
@@ -1272,6 +1346,8 @@ private void dropTableLocally(long causalityToken, CatalogTableDescriptor tableD
 
             CompletableFuture<?>[] stopReplicaFutures = new CompletableFuture<?>[partitions];
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:1347]: Seems that proposed changes are too complicated. Let's step back for a while and discuss (and design) an expected flow.

- Should we treat initial table creation as a rebalance from null assignments to pending-initial one? Or we should consider it as direct assignments.stable management? 
- Do we treat assignments.stable handling differently while it's the initial set or assignments.switch? Is it reasonable to split given handling into two methods or merge them within one?

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 12]
```diff
@@ -465,19 +464,19 @@ public static CompletableFuture<Set<Assignment>> partitionAssignments(
...
      *
-     * @param vaultManager Vault manager.
+     * @param metaStorageManager Meta storage manager.
      * @param tableId Table id.
      * @param partitionNumber Partition number.
-     * @return Returns partition assignments from vault or {@code null} if assignments is absent.
+     * @param revision Revision.
+     * @return Returns partition assignments from meta storage locally or {@code null} if assignments is absent.
      */
     public static Set<Assignment> partitionAssignments(
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:475]: fixed

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 13]
```diff
@@ -465,19 +464,19 @@ public static CompletableFuture<Set<Assignment>> partitionAssignments(
...
-            VaultManager vaultManager, int tableId, int partitionNumber) {
-        VaultEntry entry =
-                vaultManager.get(stablePartAssignmentsKey(new TablePartitionId(tableId, partitionNumber))).join();
+            MetaStorageManager metaStorageManager, int tableId, int partitionNumber, long revision) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:476]: Otherwise it would break string length limitation. I separated them into different lines.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 14]
```diff
@@ -465,19 +464,19 @@ public static CompletableFuture<Set<Assignment>> partitionAssignments(
...
+        Entry entry = metaStorageManager.getLocally(stablePartAssignmentsKey(new TablePartitionId(tableId, partitionNumber)), revision);
 
-        return (entry == null) ? null : ByteUtils.fromBytes(entry.value());
+        return (entry == null || entry.empty() || entry.tombstone()) ? null : ByteUtils.fromBytes(entry.value());
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:479]: fixed

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java [hunk 15]
```diff
@@ -516,25 +515,28 @@ static CompletableFuture<List<Set<Assignment>>> tableAssignments(
...
-                .mapToObj(i ->
-                        (Set<Assignment>) ByteUtils.fromBytes(
-                                vaultManager.get(
-                                        stablePartAssignmentsKey(new TablePartitionId(tableId, i))
-                                ).join().value())
-                )
+                .mapToObj(p -> {
+                    Entry e = metaStorageManager.getLocally(stablePartAssignmentsKey(new TablePartitionId(tableId, p)), revision);
+
+                    assert !e.empty() && !e.tombstone() : e;
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java:536]: Because we first check the presence of assignments for one partition, and then presume that others are also present. If some partitions missing, this is some kind of corruption and we throw assertion.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 8 above)
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:523]: Probably it should be renamed to `processAssignmentsOnRecovery`. We will need to process both kinds of assignments anyway.

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java (hunk 7 above)
Comment (denis-chudov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/rebalance/ItRebalanceDistributedTest.java:504]: On this step I mostly care not to break something, but probably there should be some tests for storages cleanup. I expect more tests to be added in [IGNITE-20993](https://issues.apache.org/jira/browse/IGNITE-20993)

File: modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java [hunk 16]
```diff
@@ -451,6 +451,14 @@ Publisher<BinaryRow> lookup(
      */
     RaftGroupService partitionRaftGroupService(int partition);
 
+    /**
+     * Returns raft group client for corresponding partition.
+     *
+     * @param partition Partition number.
+     * @return Whether raft group client for corresponding partition is started.
+     */
+    boolean partitionRaftGroupServiceStarted(int partition);
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java:460]: It is needed because `partitionRaftGroupService` doesn't return null if service is not started, it throws exception.

[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java:460]: Yep, but why we need it? I mean it's used in order to check whether it's required to start raft server on handling changePendingAssignmentEvent: if there's server up and running - no need to start it. But why not to compare assignments.stable and assignments.pending and start (assignments.pending - assignments.stable)?

As far as I remember we've discussed similar topic several months ago with @kgusakov thus he may have some concerns here.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/RebalanceUtil.java]: I didn't get the point. Why we treat assignments presence differently in `tableAssignmentsGetLocally` and `tableAssignments`?

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+    private WaitForTimes(long wantedTimes, long timeoutMillis) {
+        this.wantedTimes = wantedTimes;
+        this.timeoutMillis = timeoutMillis;
+    }
+
+    @Override
+    public void verify(VerificationData data) {
+        try {
+            waitForCondition(() -> {
```
Comment (sanpwc) [inline modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java:43]: Well, let's skip this for now, but proposed approach doesn't look reliable enough. We may be lucky enough to match an expected number within specific iteration. In that case we will stop the check, despite the fact that the number might have increased with further waiting.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+                MatchableInvocation wanted = data.getTarget();
+                long actual = data.getAllInvocations().stream().filter(wanted::matches).count();
+
+                if (actual > wantedTimes) {
+                    throw new AssertionError("Called more times than expected: " + actual);
```
Comment (sanpwc) [inline modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java:48]: Should we also print an expected number?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 8 above)
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:528]: Seems that it's not true for now. On node restart we handle stable assignments directly with `startLocalPartitionsAndUpdateClients` without calling `handleAssignmentsOnRecovery`.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 19]
```diff
@@ -526,39 +524,51 @@ public void start() {
...
+        startVv.update(recoveryRevision, (v, e) -> handleAssignmentsOnRecovery(
+                pendingAssignmentsPrefix,
+                recoveryRevision,
+                (entry, rev) -> handleChangePendingAssignmentEvent(entry, rev, true),
+                "pending"
+        ));
+    }
 
-        try (Cursor<Entry> cursor = metaStorageMgr.prefixLocally(prefix, recoveryRevision)) {
+    private CompletableFuture<Void> handleAssignmentsOnRecovery(
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:538]: As mentioned above `processAssignmentsOnRecovery` is only used to handle pending assignments and not the stable one. Thus either approach or naming is misleading.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 20]
```diff
@@ -612,223 +622,191 @@ private CompletableFuture<Void> onTableDelete(DropTableEventParameters parameter
...
      * Updates or creates partition raft groups and storages.
      *
-     * @param causalityToken Causality token.
      * @param assignmentsFuture Table assignments.
      * @param table Initialized table entity.
      * @return future, which will be completed when the partitions creations done.
      */
-    private CompletableFuture<?> createTablePartitionsLocally(
-            long causalityToken,
+    private CompletableFuture<Void> startLocalPartitionsAndUpdateClients(
```
Comment (sanpwc) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:629]: Where do we actually update clients? And where do we actually use that updated clients?

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java (hunk 17 above)
Comment (denis-chudov) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java:43]: well, after thh las rework this change is no longer needed, so I remove this class completely

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java (hunk 18 above)
Comment (denis-chudov) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/verification/WaitForTimes.java:48]: removed

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 20 above)
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:629]: Inside of startPartitionAndStartClient, in thenComposeAsync for startGroupFut. After that they are put into InternalTableImpl
`((InternalTableImpl) internalTbl).updateInternalTableRaftGroupService(partId, updatedRaftGroupService);`

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 19 above)
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:538]: I think it will be reasonable after stable assignments will be processed on recovery as well, so I decided to leave it as is

Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:531]: Ok

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 20 above)
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:622]: NOTSUREIF

Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:622]: Makes sense.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java [hunk 21]
```diff
@@ -1191,6 +1207,90 @@ public void updateClusterCfgWithDefaultValue() {
...
+            Set<Assignment> assignment = ByteUtils.fromBytes(e.value());
+
+            boolean shouldBe = assignment.stream().anyMatch(n -> n.consistentId().equals(restartedNode.name()));
+
+            PeersAndLearners configuration = PeersAndLearners.fromConsistentIds(assignment.stream().map(a -> a.consistentId()).collect(
+                    Collectors.toSet()), Set.of());
+
+            Peer peer = configuration.peer(restartedNode.name());
+
+            boolean isStarted = peer == null ? false : restartedNode.raftManager().isStarted(new RaftNodeId(tablePartitionId, peer));
```
Comment (kgusakov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java:1288]: Could you explain this check? As I understand - we received assignments from local metastore and check double times (with the different ways, but anyway), if this assignments has the restarted node...

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java [hunk 22]
```diff
@@ -1988,51 +1981,94 @@ protected CompletableFuture<Void> handleChangeStableAssignmentEvent(WatchEvent e
...
+                            pendingAssignments,
+                            isRecovery,
+                            revision
+                    );
+                }, ioExecutor);
+    }
 
-                    boolean shouldStopLocalServices = Stream.concat(stableAssignments.stream(), pendingAssignments.stream())
-                            .noneMatch(assignment -> assignment.consistentId().equals(localMemberName));
+    private CompletableFuture<Void> updateClientsOnPartitionStop(
```
Comment (kgusakov) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2032]: It is hard to understand, why we need to update the clients on partition stop. Why do you want to move it from the handleChangeStableAssignmentEvent to stopAndDestroyPartitionAndUpdateClients?

**Review by kgusakov — CHANGES_REQUESTED** [review]

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 22 above)
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2032]: Because it appears that we should only stop partitions on handling of stable assignments. I decided to move this part to separate method because it improved the readability of code, as for me.

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java (hunk 21 above)
Comment (denis-chudov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java]: The reason is that I've actually taken this part from table manager, and I didn't like this check and thought it would be nice to create some fake configuration in test that includes all nodes, to be able to create raft node id. I even thought about it but forgot in final version.

Could you check the current version?

File: modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java:460]: Any updates?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java (hunk 22 above)
Comment (kgusakov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2009]: Ok, but the method itself is not connected with stop, it is just updating the clients. maybe the "updatePartitoinClients" will be better?

Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java:2009]: ok will rename it

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java (hunk 21 above)
Comment (kgusakov) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/ItIgniteNodeRestartTest.java]: Yep, thanks!

File: modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/InternalTable.java]: I think I can't remember the exact issue which was the reason why this method was needed, but it's gone after the latest rework. I removed this method.

**Review by sanpwc — APPROVED** [review]

**Review by kgusakov — APPROVED** [review]
